cardano-api/src/Cardano/Api/Convenience/Construction.hs

```diff
@@ -0,0 +1,102 @@
+-- | Convenience transaction construction functions
+--
+module Cardano.Api.Convenience.Construction (
+    constructBalancedTx,
+
+    -- * Misc
+    TxInsExistError(..),
+    NotScriptLockedTxInsError(..),
+    notScriptLockedTxIns,
+    renderNotScriptLockedTxInsError,
+    renderTxInsExistError,
+    txinsExist,
+
+  ) where
+
+import           Prelude
+
+import qualified Data.List as List
+import qualified Data.Map.Strict as Map
+import           Data.Set (Set)
+import qualified Data.Set as Set
+import           Data.Text (Text)
+import qualified Data.Text as Text
+
+import           Cardano.Api.Address
+import           Cardano.Api.Certificate
+import           Cardano.Api.Eras
+import           Cardano.Api.Fees
+import           Cardano.Api.IPC
+import           Cardano.Api.Modes
+import           Cardano.Api.ProtocolParameters
+import           Cardano.Api.Query
+import           Cardano.Api.Tx
+import           Cardano.Api.TxBody
+import           Cardano.Api.TxSubmit.Types (textShow)
+
+
+-- | Construct a balanced transaction.
+-- See Cardano.Api.Convenience.Query.queryStateForBalancedTx for a
+-- convenient way of querying the node to get the required arguements
+-- for constructBalancedTx.
+constructBalancedTx
+  :: IsShelleyBasedEra era
+  => EraInMode era CardanoMode
+  -> TxBodyContent BuildTx era
+  -> AddressInEra era -- ^ Change address
+  -> Maybe Word       -- ^ Witness override
+  -> UTxO era
+  -> ProtocolParameters
+  -> EraHistory CardanoMode
+  -> SystemStart
+  -> Set PoolId
+  -> [KeyWitness era]
+  -> Either TxBodyErrorAutoBalance (Tx era)
+constructBalancedTx eInMode txbodcontent changeAddr mOverrideWits utxo pparams
+                    eraHistory systemStart stakePools keyWits = do
+  BalancedTxBody txbody _txBalanceOutput _fee
+    <- makeTransactionBodyAutoBalance
+         eInMode systemStart eraHistory
+         pparams stakePools utxo txbodcontent
+         changeAddr mOverrideWits
+
+  return $ makeSignedTransaction keyWits txbody
+
+data TxInsExistError
+  = TxInsDoNotExist [TxIn]
+  | EmptyUTxO
+
+renderTxInsExistError :: TxInsExistError -> Text
+renderTxInsExistError EmptyUTxO =
+  "The UTxO is empty"
+renderTxInsExistError (TxInsDoNotExist txins) =
+  "The following tx input(s) were not present in the UTxO: " <>
+  Text.singleton '\n' <>
+  Text.intercalate (Text.singleton '\n') (map renderTxIn txins)
+
+txinsExist :: [TxIn] -> UTxO era -> Either TxInsExistError ()
+txinsExist ins (UTxO utxo)
+  | null utxo = Left EmptyUTxO
+  | otherwise = do
+      let utxoIns = Map.keys utxo
+          occursInUtxo = [ txin | txin <- ins, txin `elem` utxoIns ]
+      if length occursInUtxo == length ins
+      then return ()
+      else Left . TxInsDoNotExist $ ins List.\\ ins `List.intersect` occursInUtxo
```

Comment:
Everything in `occursInUtxo` is drawn from `ins` to begin with, so intersecting with `ins` is identity (and therefore a waste of cycles)

```suggestion
      else Left . TxInsDoNotExist $ ins List.\\ occursInUtxo
```

Reply:
good catch :+1: